public/app/plugins/datasource/loki/querybuilder/components/QueryPatternsModal.tsx

```diff
@@ -0,0 +1,176 @@
+
+import logqlGrammar from '../../syntax';
+import { LokiQuery } from '../../types';
+import { lokiQueryModeller } from '../LokiQueryModeller';
+import { buildVisualQueryFromString } from '../parsing';
+import { LokiQueryPattern, LokiQueryPatternType } from '../types';
+
+type Props = {
+  isOpen: boolean;
+  query: LokiQuery;
+  queries: DataQuery[] | undefined;
+  onClose: () => void;
+  onChange: (query: LokiQuery) => void;
+  onAddQuery?: (query: LokiQuery) => void;
+};
+
+export const QueryPatternsModal = (props: Props) => {
+  const { isOpen, onClose, onChange, onAddQuery, query, queries } = props;
+  const [openTabs, setOpenTabs] = useState<string[]>([]);
+  const [selectedPatternConfirmationIndex, setSelectedPatternConfirmationIndex] = useState<number | null>(null);
+  const indexOfSelectedCard = useRef<number | null>(null);
+
+  const styles = useStyles2(getStyles);
+  const lang = { grammar: logqlGrammar, name: 'logql' };
+  const hasPreviousQuery = buildVisualQueryFromString(query.expr).query.operations.length > 0;
+  const hasNewQueryOption = !!onAddQuery;
+
+  const onPatternSelect = (pattern: LokiQueryPattern, selectAsNewQuery = false) => {
+    const queryString = selectAsNewQuery ? '' : query.expr;
+    const visualQuery = buildVisualQueryFromString(queryString);
+    visualQuery.query.operations = pattern.operations;
+
+    if (hasNewQueryOption && selectAsNewQuery) {
+      onAddQuery({
+        ...query,
+        refId: getNextRefIdChar(queries ?? [query]),
+        expr: lokiQueryModeller.renderQuery(visualQuery.query),
+      });
+    } else {
+      onChange({
+        ...query,
+        expr: lokiQueryModeller.renderQuery(visualQuery.query),
+      });
+    }
+    setSelectedPatternConfirmationIndex(null);
+    onClose();
+  };
+
+  return (
+    <Modal isOpen={isOpen} title="Kick start your query" onDismiss={onClose}>
+      <div className={styles.spacing}>
+        Kick start your query by selecting one of these queries. You can then continue to complete your query.
+      </div>
+      {Object.values(LokiQueryPatternType).map((patternType) => {
+        return (
+          <Collapse
+            key={patternType}
+            label={`${patternType[0].toUpperCase() + patternType.slice(1)} query starters`}
+            isOpen={openTabs.includes(patternType)}
+            collapsible={true}
+            onToggle={() =>
+              setOpenTabs((tabs) =>
+                // close tab if it's already open, otherwise open it
+                tabs.includes(patternType) ? tabs.filter((t) => t !== patternType) : [...tabs, patternType]
+              )
+            }
+          >
+            <div className={styles.cardsContainer}>
+              {lokiQueryModeller
+                .getQueryPatterns()
+                .filter((pattern) => pattern.type === patternType)
+                .map((pattern, index) => (
+                  <Card key={pattern.name} className={styles.card}>
```

Comment:
I would consider moving this whole `<Card>` block to it's own component. E.g. a `<QueryPattern>` component.

Reply:
Yes, I was thinking about it as well! Good suggestion, I'll do that.

Reply:
Updated in https://github.com/grafana/grafana/pull/55097/commits/c3ec9cc7af2d9caa8535f7c6c6405129a3c53497. It is much cleaner and understandable. I also discovered a bug related to the fact, that I was relying on index rather than pattern name. And also I had leftover `ref` that I wasn't using anymore.